src/components/join-flow/join-flow.jsx

```diff
@@ -61,59 +64,111 @@ class JoinFlow extends React.Component {
+    // recaptcha messages:
+    //   * "recaptcha": ["This field is required."]
+    //   * "recaptcha": ["Incorrect, please try again."]
+    //   * "recaptcha": [some timeout message?]
+    // other messages:
+    //   * "birth_month": ["Ensure this value is less than or equal to 12."]
+    //   * "birth_month": ["Ensure this value is greater than or equal to 1."]
     handleRegistrationResponse (err, body, res) {
-        // example of failing response:
-        // [
-        //   {
-        //     "msg": "This field is required.",
-        //     "errors": {
-        //       "username": ["This field is required."],
-        //       "recaptcha": ["Incorrect, please try again."]
-        //     },
-        //     "success": false
-        //   }
-        // ]
-        // username: 'username exists'
         this.setState({
             numAttempts: this.state.numAttempts + 1,
             waiting: false
         }, () => {
-            let errStr = '';
-            if (!err && res.statusCode === 200) {
-                if (body && body[0]) {
-                    if (body[0].success) {
-                        this.props.refreshSession();
-                        this.setState({
-                            step: this.state.step + 1
-                        });
-                        return;
-                    }
-                    if (body[0].errors) {
-                        // body can include zero or more error objects, each
-                        // with its own key and description. Here we assemble
-                        // all of them into a single string, errStr.
-                        const errorKeys = Object.keys(body[0].errors);
-                        errorKeys.forEach(key => {
-                            const val = body[0].errors[key];
-                            if (val && val[0]) {
-                                if (errStr.length) errStr += '; ';
-                                errStr += `${key}: ${val[0]}`;
-                            }
-                        });
-                    }
-                    if (!errStr.length && body[0].msg) errStr = body[0].msg;
+            const success = this.getRegistrationSuccess(err, body, res);
+            if (success) {
+                this.props.refreshSession();
+                this.setState({step: this.state.step + 1});
+                return;
+            }
+            // now we know there was some error.
+            // if the error was client-side, prompt user to try again
+            if (err || (res.statusCode >= 400 && res.statusCode < 500)) {
+                this.setState({registrationError: {canTryAgain: true}});
+                return;
+            }
+            // now we know error was server-side.
+            // if server provided us info on why registration failed,
+            // build a summary explanation string
+            let errorMsg = null;
+            const bodyErrors = this.getBodyErrors(err, body, res);
```

Comment:
if getBodyErrors instead returns more understandable object, then we can just say if (length === 1) look up single error str...

Reply:
Done